File: sockets-lesson2/web-server.py
### adapted from https://github.com/realpython/materials/tree/master/python-sockets-tutorial
## and from https://stackoverflow.com/questions/21153262/sending-html-through-python-socket-server

import socket

HOST = "127.0.0.1"  # Standard loopback interface address (localhost)
port_number = 65433  # Port to listen on (non-privileged ports are > 1023)

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as aSocket:
    aSocket.bind((HOST, port_number))
    aSocket.listen()
    print("server started at address %s on port %i" % (HOST, port_number))
    connection, address = aSocket.accept()
    with connection:
        print("Client connected from: ", address)
        while True:
            data = connection.recv(1024)
            if not data:
                break
            print("message from client: \n", data.decode())
            print(" -- end message from client -- \n")
            connection.send('HTTP/1.0 200 OK\n'.encode())
            connection.send('Content-Type: text/html\n'.encode())
            connection.send('\n'.encode()) # header and body should be separated by additional newline
            connection.send("""
                <html>
                <body>
                <h1>Hello World</h1> this is my server, damnit!
                </body>
                </html>
                """.encode())  # Use triple-quote string.
            connection.close()

File: archive/http-intro-js/web-server.js
const net = require('net');

const HOST = '0.0.0.0'; // Standard loopback interface address (localhost)
const port_number = 65432; // Port to listen on (non-privileged ports are > 1023)

const server = net.createServer((socket) => {
  console.log(`Server started at address ${HOST} on port ${port_number}`);
  socket.on('data', (data) => {
    console.log('Client connected from: ', socket.remoteAddress, socket.remotePort);
    const requestData = data.toString();

    // Log received message from client
    console.log('Message from client:\n', requestData);
    console.log(' -- end message from client -- ');

    // HTTP Response
    const httpStatusLine = 'HTTP/1.0 200 OK';
    const httpHeaders = 'Content-Type: text/html\r\nContent-Length: 44'; // Modify Content-Length accordingly
    const httpBody = `
<html>
<body>
<h1>Hello World</h1> 
<p>This is my server, hey.</p>
</body>
</html>
`;

    const httpResponse = `${httpStatusLine}\r\n${httpHeaders}\r\n\r\n${httpBody}`;
    socket.write(httpResponse, 'utf-8', () => {
      // Close the connection after sending the response
      socket.end();
      console.log('Connection closed');
    });
  });

  socket.on('error', (err) => {
    console.error('Socket error:', err);
  });
});

server.listen(port_number, HOST);